pkg/controller/atlasbackupcompliancepolicy/atlasbackupcompliancepolicy_controller.go

```diff
@@ -0,0 +1,202 @@
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api"
+	akov2 "github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api/v1"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/controller/atlas"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/controller/customresource"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/controller/statushandler"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/controller/workflow"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/indexer"
+)
+
+type AtlasBackupCompliancePolicyReconciler struct {
+	Client                      client.Client
+	Scheme                      *runtime.Scheme
+	GlobalPredicates            []predicate.Predicate
+	EventRecorder               record.EventRecorder
+	AtlasProvider               atlas.Provider
+	Log                         *zap.SugaredLogger
+	ObjectDeletionProtection    bool
+	SubObjectDeletionProtection bool
+}
+
+//TODO kubebuilder rbac
+
+func (r *AtlasBackupCompliancePolicyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.With("atlasbackupcompliancepolicy", req.NamespacedName)
+	log.Infow("-> Starting AtlasBackupCompliancePolicy reonciliation")
+
+	bcp := &akov2.AtlasBackupCompliancePolicy{}
+	result := customresource.PrepareResource(ctx, r.Client, req, bcp, log)
+	if !result.IsOk() {
+		return result.ReconcileResult(), nil
+	}
+
+	if customresource.ReconciliationShouldBeSkipped(bcp) {
+		return r.skip(ctx, log, bcp), nil
+	}
+
+	conditions := akov2.InitCondition(bcp, api.FalseCondition(api.ReadyType))
+	workflowCtx := workflow.NewContext(log, conditions, ctx)
+	defer statushandler.Update(workflowCtx, r.Client, r.EventRecorder, bcp)
+
+	isValid := customresource.ValidateResourceVersion(workflowCtx, bcp, r.Log)
+	if !isValid.IsOk() {
+		return r.invalidate(isValid)
+	}
+
+	if !r.AtlasProvider.IsResourceSupported(bcp) {
+		return r.unsupport(workflowCtx)
+	}
+
+	return r.ensureAtlasBackupCompliancePolicy(workflowCtx, bcp)
+}
+
+func (r *AtlasBackupCompliancePolicyReconciler) ensureAtlasBackupCompliancePolicy(workflowCtx *workflow.Context, bcp *akov2.AtlasBackupCompliancePolicy) (ctrl.Result, error) {
+	projects := &akov2.AtlasProjectList{}
+	listOpts := &client.ListOptions{
+		FieldSelector: fields.OneTermEqualSelector(
+			indexer.AtlasProjectByBackupCompliancePolicyIndex,
+			client.ObjectKeyFromObject(bcp).String(),
+		),
+	}
+	err := r.Client.List(workflowCtx.Context, projects, listOpts)
+	if err != nil {
+		return r.terminate(workflowCtx, workflow.Internal, err)
+	}
+
+	if len(projects.Items) > 0 {
+		return r.lock(workflowCtx, bcp)
+	}
+
+	return r.release(workflowCtx, bcp)
```

Comment:
This is the gist of the backup policy state machine here right?
It is just that simple, basically locking so long as there are references.

Reply:
This controller solely manages the finalisers (i.e. whether or not projects reference a BCP). The rest of the logic (creation, updating, etc.) happens in the project controller, as the BCP is actually just settings on a project.

Reply:
@josvazg yes, this is the same here as streams processing.